integration/registry_docker_config.go

```diff
@@ -0,0 +1,104 @@
+package integration
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+	"os/exec"
+	"testing"
+
+	"github.com/sclevine/spec"
+	"github.com/stretchr/testify/require"
+)
+
+var _ = suite("registry/docker-config", func(t *testing.T, when spec.G, it spec.S) {
+	var (
+		expect          *require.Assertions
+		server          *httptest.Server
+		expectReadWrite bool
+	)
+
+	it.Before(func() {
+		expect = require.New(t)
+
+		server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			w.Header().Add("content-type", "application/json")
+
+			switch req.URL.Path {
+			case "/v2/registry":
+				auth := req.Header.Get("Authorization")
+				if auth != "Bearer some-magic-token" {
+					w.WriteHeader(http.StatusUnauthorized)
+					return
+				}
+
+				if req.Method != http.MethodGet {
+					w.WriteHeader(http.StatusMethodNotAllowed)
+					return
+				}
+
+				w.Write([]byte(registryGetResponse))
+			case "/v2/registry/docker-credentials":
+				auth := req.Header.Get("Authorization")
+				if auth != "Bearer some-magic-token" {
+					w.WriteHeader(http.StatusUnauthorized)
+					return
+				}
+
+				if req.Method != http.MethodGet {
+					w.WriteHeader(http.StatusMethodNotAllowed)
+					return
+				}
+
+				readWriteParam := req.URL.Query().Get("read_write")
+				if expectReadWrite {
+					expect.Contains([]string{"true", "1"}, readWriteParam)
+				} else {
+					expect.Contains([]string{"false", "", "0"}, readWriteParam)
+				}
+
+				w.Write([]byte(registryDockerCredentialsResponse))
+			default:
+				dump, err := httputil.DumpRequest(req, true)
+				if err != nil {
+					t.Fatal("failed to dump request")
+				}
+
+				t.Fatalf("received unknown request: %s", dump)
+			}
+		}))
+	})
+
+	it("prints the returned read-only docker config", func() {
+		expectReadWrite = false
```

Comment:
Is this safe? And, is there a better way to test both the read-only and read/write cases? (This doesn't feel great, but since the expectations are set up in the `Before` it wasn't clear there was a nicer way to go.)

Reply:
I think generally we don't want the test behavior to vary based on something defined outside of the command that is being called. I might have the server do something like this:

```
readWriteParam := req.URL.Query().Get("read_write")
if readWriteParam != "true" {
	w.Write([]byte(registryDockerCredentialsReadOnlyResponse))
} else {
	w.Write([]byte(registryDockerCredentialsReadWriteResponse))
}
```

The contents of the auth string in the response bodies aren't particularly important.

Reply:
Good idea - done.